README.md
<h1> streamlit-annotation </h1>

- [Streamlit Annotation Tools](#streamlit-annotation-tools)
- [Demo](#demo)
  - [Try](#try)
  - [Text Highlighter](#text-highlighter)
  - [Text Labeler](#text-labeler)
- [Install](#install)
- [Quick Use](#quick-use)
  - [Text Highlighter](#text-highlighter-1)
  - [Text Labeler](#text-labeler-1)
- [Development](#development)
  - [Install](#install-1)
  - [Run](#run)
- [License](#license)
- [Author](#author)
- [Contributors](#contributors)

# Streamlit Annotation Tools

[![PyPI version](https://badge.fury.io/py/streamlit-annotation-tools.svg)](https://badge.fury.io/py/streamlit-annotation-tools)
[![Downloads](https://pepy.tech/badge/streamlit-annotation-tools)](https://pepy.tech/project/streamlit-annotation-tools)
[![Downloads](https://pepy.tech/badge/streamlit-annotation-tools/month)](https://pepy.tech/project/streamlit-annotation-tools)
[![Downloads](https://pepy.tech/badge/streamlit-annotation-tools/week)](https://pepy.tech/project/streamlit-annotation-tools)

Streamlit Annotation Tools is a Streamlit component that gives you access to various annotation tools (labeling, annotation, etc.) for text data.

# Demo

## Try
[![Open in Streamlit](https://static.streamlit.io/badges/streamlit_badge_black_white.svg)](https://annotation-tools.streamlit.app/)

## Text Highlighter
![](docs/highlight_tool.gif)

## Text Labeler
![](docs/label_tool.gif)

# Install

```
pip install streamlit-annotation-tools
```

# Quick Use

## Text Highlighter

Create an example.py file

```python
from streamlit_annotation_tools import text_highlighter

text = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean et hendrerit orci. Praesent auctor malesuada lobortis. Suspendisse ac elit bibendum, congue tellus vel, ornare ipsum. Mauris at tellus in turpis aliquet cursus."

annotations = text_highlighter(text)
```

Run:

```
streamlit run example.py
```

Output:

```python
[
  [
    {"start": 0, "end": 5, "label": "Lorem"},
    {"start": 12, "end": 17, "label": "dolor"},
  ],
  [
    {"start": 6, "end": 11, "label": "ipsum"},
    {"start": 18, "end": 21, "label": "sit"},
  ],
]
```

## Text Labeler

Create an example.py file

```python
from streamlit_annotation_tools import text_labeler

text = "Yesterday, at 3 PM, Emily Johnson and Michael Smith met at the Central Park in New York to discuss the merger between TechCorp and Global Solutions. The deal, worth approximately 500 million dollars, is expected to significantly impact the tech industry. Later, at 6 PM, they joined a conference call with the CEO of TechCorp, David Brown, who was in London for a technology summit. During the call, they discussed the market trends in Asia and Europe and planned for the next quarterly meeting, which is scheduled for January 15th, 2024, in Paris."

labels = text_labeler(text)
```

Run:

```
streamlit run example.py
```

Output:

```python
{
    "Personal names": [
        {"start": 20, "end": 33, "label": "Emily Johnson"},
        {"start": 38, "end": 51, "label": "Michael Smith"},
        {"start": 327, "end": 338, "label": "David Brown"},
    ],
    "Organizations": [
        {"start": 118, "end": 126, "label": "TechCorp"},
        {"start": 131, "end": 147, "label": "Global Solutions"},
    ],
    "Locations": [
        {"start": 63, "end": 75, "label": "Central Park"},
        {"start": 79, "end": 87, "label": "New York"},
        {"start": 351, "end": 357, "label": "London"},
        {"start": 436, "end": 440, "label": "Asia"},
        {"start": 445, "end": 451, "label": "Europe"},
        {"start": 542, "end": 547, "label": "Paris"},
    ],
    "Time": [
        {"start": 0, "end": 9, "label": "Yesterday"},
        {"start": 14, "end": 18, "label": "3 PM"},
        {"start": 265, "end": 269, "label": "6 PM"},
        {"start": 519, "end": 531, "label": "January 15th"},
        {"start": 533, "end": 537, "label": "2024"},
    ],
    "Money": [{"start": 179, "end": 198, "label": "500 million dollars"}],
}
```

# Development

## Install

```
git clone git@github.com:rmarquet21/streamlit-annotation-tools.git
cd streamlit-annotation-tools
pip install -e .
```

## Run

```
streamlit run example.py
```

# License

MIT

# Author

Robin Marquet

# Contributors

- [Robin Marquet](robin.marquet3@gmail.com)



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
## [Unreleased] - yyyy-mm-dd
 
Here we write upgrading notes for brands. It's a team effort to make them as
straightforward as possible.
 
### Added
 
### Changed
 
### Fixed

## [1.0.1] - 2023-12-09

### Fixed

- Support for Streamlit 1.21.0 minimum [#3](https://github.com/rmarquet21/streamlit-annotation-tools/issues/3)

## [1.0.0] - 2023-12-08

### Added

- Add `text_labeler` component

### Changed

- rename 'text_annotation' to 'text_highlighter'

## [0.3.4] - 2023-12-07

### Changed

- Update architecture of frontend components

### Fixed

- hidden state when adding a new list [#1](https://github.com/rmarquet21/streamlit-annotation-tools/issues/1)

## [0.3.3] - 2023-05-10

### Fixed

- Special characters are not removed anymore

## [0.3.1] - 2023-05-10

### Fixed

- Impossible to select a punctuation at the end of an annotation, it is directly removed from the final annotation. (Now to select with punctuation at the end, you have to highlight until the final punctuation, without letting the autocompletion select the entire last word)

## [0.3.0] - 2023-05-08

### Added

- New UI

### Changed

- Update to latest version of `React v18`

### Fixed

- Text dispear on hover

## [0.2.2] - 2023-04-17

### Fixed

- Fix import of build

## [0.2.0] - 2023-04-17
 
### Added

- Initial release


src/streamlit_annotation_tools/__init__.py
import os

import streamlit.components.v1 as components

RELEASE = os.environ.get("STREAMLIT_ANNOTATION_TOOLS_RELEASE", "true") == "true"

if not RELEASE:
    _component = components.declare_component(
        "annotation_tools", url="http://localhost:3001"
    )
else:
    parent_dir = os.path.dirname(os.path.abspath(__file__))
    build_dir = os.path.join(parent_dir, "frontend/build")
    _component = components.declare_component("annotation_tools", path=build_dir)


def text_highlighter(text: str, highlights=[]):
    """Create a new instance of "text_highlighter".

    Parameters
    ----------
    text : str
        The text to be annotated

    highlights: list
        If the text has already been annotated, the highlights can be passed to the component in the form of a list of dictionaries with the following structure:
        [
            [
                {
                    label: "label",
                    start: 0,
                    end: 10
                },
                {
                    label: "label",
                    start: 0,
                    end: 10
                }
            ],
            [
                {
                    label: "label",
                    start: 0,
                    end: 10
                }
            ]
        ]

    Returns
    -------
    list or None
        The highlights made by the user in the form of a list of dictionaries with the following structure:
        [
            [
                {
                    label: "label",
                    start: 0,
                    end: 10
                },
                {
                    label: "label",
                    start: 0,
                    end: 10
                }
            ],
            [
                {
                    label: "label",
                    start: 0,
                    end: 10
                }
            ]
        ]
    """

    # Call through to our private component function. Arguments we pass here
    # will be sent to the frontend, where they'll be available in an "args"
    # dictionary.
    #
    # "default" is a special argument that specifies the initial return
    # value of the component before the user has interacted with it.
    component_value = _component(
        text=text, highlights=highlights, mode="text_highlighter"
    )

    # We could modify the value returned from the component if we wanted.
    # There's no need to do this in our simple example - but it's an option.
    return component_value


def text_labeler(text: str, labels=[], in_snake_case=False):
    """Create a new instance of "text_labeler".

    Parameters
    ----------
    text : str
        The text to be labeled

    labels : list
        If the text has already been labeled, the labels can be passed to the component in the form of a list of dictionaries with the following structure:
        {
            "label1": [
                {
                    label: "label",
                    start: 0,
                    end: 10
                },
                {
                    label: "label",
                    start: 0,
                    end: 10
                }
            ],
        }

    in_snake_case : bool
        If True, the labels will be converted to snake case before being returned.

    Returns
    -------
    list or None
        The labels made by the user in the form of a list of dictionaries with the following structure:
        {
            "label1": [
                {
                    label: "label",
                    start: 0,
                    end: 10
                },
                {
                    label: "label",
                    start: 0,
                    end: 10
                }
            ],
        }
    """

    # Call through to our private component function. Arguments we pass here
    # will be sent to the frontend, where they'll be available in an "args"
    # dictionary.
    #
    # "default" is a special argument that specifies the initial return
    # value of the component before the user has interacted with it.
    component_value = _component(
        text=text, labels=labels, in_snake_case=in_snake_case, mode="text_labeler"
    )

    # We could modify the value returned from the component if we wanted.
    # There's no need to do this in our simple example - but it's an option.
    return component_value


if __name__ == "__main__":
    # This is an example of how to use the component in a Streamlit app.
    # It's not required for the component to work.
    import streamlit as st

    def highlighter_page():
        st.title("Text Annotation Tool")

        text = "Lorem ipsum dolor sit et amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."

        highlights = [
            [
                {"start": 0, "end": 5, "label": "Lorem"},
                {"start": 12, "end": 17, "label": "dolor"},
            ],
            [
                {"start": 6, "end": 11, "label": "ipsum"},
                {"start": 18, "end": 21, "label": "sit"},
            ],
        ]

        highlights = text_highlighter(text, highlights)

        st.write("Annotations:")
        st.write(highlights)

    def labeler_page():
        st.title("Text Labeling Tool")

        text = "Yesterday, at 3 PM, Emily Johnson and Michael Smith met at the Central Park in New York to discuss the merger between TechCorp and Global Solutions. The deal, worth approximately 500 million dollars, is expected to significantly impact the tech industry. Later, at 6 PM, they joined a conference call with the CEO of TechCorp, David Brown, who was in London for a technology summit. During the call, they discussed the market trends in Asia and Europe and planned for the next quarterly meeting, which is scheduled for January 15th, 2024, in Paris."

        labels = {
            "Personal names": [
                {"start": 20, "end": 33, "label": "Emily Johnson"},
                {"start": 38, "end": 51, "label": "Michael Smith"},
                {"start": 327, "end": 338, "label": "David Brown"},
            ],
            "Organizations": [
                {"start": 118, "end": 126, "label": "TechCorp"},
                {"start": 131, "end": 147, "label": "Global Solutions"},
            ],
            "Locations": [
                {"start": 63, "end": 75, "label": "Central Park"},
                {"start": 79, "end": 87, "label": "New York"},
                {"start": 351, "end": 357, "label": "London"},
                {"start": 436, "end": 440, "label": "Asia"},
                {"start": 445, "end": 451, "label": "Europe"},
                {"start": 542, "end": 547, "label": "Paris"},
            ],
            "Time": [
                {"start": 0, "end": 9, "label": "Yesterday"},
                {"start": 14, "end": 18, "label": "3 PM"},
                {"start": 265, "end": 269, "label": "6 PM"},
                {"start": 519, "end": 531, "label": "January 15th"},
                {"start": 533, "end": 537, "label": "2024"},
            ],
            "Money": [{"start": 179, "end": 198, "label": "500 million dollars"}],
        }

        labels = text_labeler(text, labels, in_snake_case=False)

        st.write("Labels:")
        st.write(labels)

    pages = {"Highlighter tool": highlighter_page, "Labeler tool": labeler_page}

    demo_name = st.sidebar.selectbox("Choose a demo", pages.keys())
    pages[demo_name]()
